### ISSUE TITLE
[Bug]: Issue with feature flag config in the Hello World section example

### ISSUE THREAD

**Issue by markbiek** [Automattic/wp-calypso#61347]
### Quick summary

While working through https://wpcalypso.wordpress.com/devdocs/docs/guide/hello-world.md I ran into the following issue.

In Step 5 where you register the new section by adding 

`const config = require('@automattic/calypso-config').default;` 

to the top of `client/sections.js` which causes the following error on `yarn start`:

> ReferenceError: No configuration was found

@noahtallen found [this section in webpack.config.js](https://github.com/Automattic/wp-calypso/blob/b57297fac00a56cf103f3235e33f6636ddca1f49/client/webpack.config.js#L230) which may be related.

### Steps to reproduce

1. Add `const config = require( '@automattic/calypso-config' ).default;` to the top of `client/sections.js`.
2. Run `yarn start`.

### What you expected to happen

The local development environment should start normally.

### What actually happened

The following error occurs:

> ReferenceError: No configuration was found

### Reproducibility

Consistent

**Comment by tyxla** [comment]
Hey @markbiek!

Thanks for bringing up the issue! Those docs haven't been updated in a while and I think it's good opportunity to iron out any inconsistencies.

The `sections.js` module is imported on the server-side, and because of that, it can't use the client-side config that's suggested in the documentation. The config data is exposed to the client by declaring a `window.configData` global variable with the configuration and feature flags. However, this isn't possible on the server side, where we have a different mechanism for providing the configuration.

In a nutshell, to fix the problem, I assume the fix should be straightforward - just replacing:

```
const config = require('@automattic/calypso-config').default;
```

with:

```
const config = require( './server/config' );
```

should do the job.

Let me know if it works or not! And if it does, I'd really appreciate it if you could file a PR to fix the obsolete docs. 

Thanks a lot 🙌

**Comment by jsnajdr** [comment]
I'm afraid that the capability to import the config module in `sections.js` has been broken in the last year or two and it will be hard to bring it back. There were refactorings that moved the config module to separate packages like `@automattic/calypso-config`, various webpack build modifications... One of them was the culprit.

The `sections.js` file is loaded into three environments:
- the bundle that runs in the browser. Here the `@automattic/calypso-config` module resolves directly to what is in `packages/calypso-config`, without any tricks. And that module provides the _browser_ side configuration code.
- the bundle that runs on the server in Node.js. Here the `@automattic/calypso-config` module is [aliased](https://github.com/Automattic/wp-calypso/blob/trunk/client/webpack.config.node.js#L138) to `calypso/server/config`, a very different module, one that provides the _server_ side configuration code.
- the webpack build process itself. The `section.js` file is processed by a `section-loader`, which [directly requires](https://github.com/Automattic/wp-calypso/blob/trunk/build-tools/webpack/sections-loader.js#L73) the source file, loading it into webpack environment, and then generates code (string) based on what it loads.

It's this third environment, webpack itself, that makes `sections.js` exceptional and that causes trouble. When required into webpack, the module tries to load `@automattic/calypso-config`, which is a browser module and it can't find the config. It should load the server config.

The solution that @tyxla proposes, to just write directly
```js
const config = require( './server/config' );
```
it will break the browser build: the `config.isEnabled` call that runs in the browser bundle will use the server config object and will fail: it attempt to do Node.js things like loading the `path` and `fs` Node modules and trying to read files with them.

There are several solutions:
If you want to just enable certain routes only in certain environment, make the route definitions conditional:
```js
if ( config.isEnabled( 'foo' ) ) {
  page( '/foo', ... );
  page( '/foo/:bar', ... );
}
```
This is used quite widely across Calypso.

We could fix the "loading config in `sections.js`" problem by rewriting the `sections-loader`. Instead of directly loading `sections.js` as JS module, it should parse it as source file and do its transformations on the AST. After all, that's what loaders usually do. The direct loading is a rather sloppy way to achieve a goal, a bad practice.

Or maybe we could swap the server and browser config modules, so that the server module is loaded "directly" and the browser module needs to be aliased? That would make things work, too. But it's my least preferred solution.

**Comment by tyxla** [comment]
Alternatively, what we often do in many of the sections is always register the section (just add it to the array of sections), but use the feature flags to decide whether to register some (or all) of its routes or not. For example: 

https://github.com/Automattic/wp-calypso/blob/52eabb4e83c28e3ed79136fab8948b9f441cf3eb/client/my-sites/marketplace/index.js#L25

**Comment by jsnajdr** [comment]
Yes, that's exactly what I meant in my "make the route definitions conditional" proposal.

**Comment by markbiek** [comment]
Thanks everyone! 

Making the call to register the route conditional definitely seems like the way to go. I'll get a PR opened for that.
